compiler/rustc_hir_typeck/src/closure.rs

```diff
@@ -470,14 +471,37 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         // for closures and async closures, respectively.
         match closure_kind {
             hir::ClosureKind::Closure
-                if self.tcx.fn_trait_kind_from_def_id(trait_def_id).is_some() => {}
+                if self.tcx.fn_trait_kind_from_def_id(trait_def_id).is_some() =>
+            {
+                self.extract_sig_from_projection(cause_span, projection)
+            }
+            hir::ClosureKind::CoroutineClosure(hir::CoroutineDesugaring::Async)
+                if self.tcx.async_fn_trait_kind_from_def_id(trait_def_id).is_some() =>
+            {
+                self.extract_sig_from_projection(cause_span, projection)
+            }
+            // It's possible we've passed the closure to a (somewhat out-of-fashion)
+            // `F: FnOnce() -> Fut, Fut: Future<Output = T>` style bound. Let's still
+            // guide inference here, since it's beneficial for the user.
             hir::ClosureKind::CoroutineClosure(hir::CoroutineDesugaring::Async)
-                if self.tcx.async_fn_trait_kind_from_def_id(trait_def_id).is_some() => {}
+                if self.tcx.fn_trait_kind_from_def_id(trait_def_id).is_some() =>
+            {
+                self.extract_sig_from_projection_and_future_bound(cause_span, projection)
+            }
             _ => return None,
         }
+    }
+
+    /// Given an `FnOnce::Output` or `AsyncFn::Output` projection, extract the args
+    /// and return type to infer a [`ty::PolyFnSig`] for the closure.
+    fn extract_sig_from_projection(
+        &self,
+        cause_span: Option<Span>,
+        projection: ty::PolyProjectionPredicate<'tcx>,
+    ) -> Option<ExpectedSig<'tcx>> {
+        let projection = self.resolve_vars_if_possible(projection);
 
         let arg_param_ty = projection.skip_binder().projection_term.args.type_at(1);
-        let arg_param_ty = self.resolve_vars_if_possible(arg_param_ty);
         debug!(?arg_param_ty);
 
         let ty::Tuple(input_tys) = *arg_param_ty.kind() else {
```

Comment:
in what case does this end up not being a tuple?

Reply:
Will we get here even though an error has been reported when `FnOnce<Args>` does not satisfy `Args: Tuple`?

Reply:
> in what case does this end up not being a tuple?

Generic code, where we'll see a type parameter and not a tuple.

> Will we get here even though an error has been reported when FnOnce<Args> does not satisfy Args: Tuple?

There will be no error reported if we have a type parameter and a `A: Tuple` bound, but there will likely be a simple type mismatch error later on:

```
fn uwu<A: Tuple>(_: impl Fn<A, Output = ()>) {}

fn test<A: Tuple>() {
  uwu::<A>(|| {});
}
```

Reply:
Resulting in:

```
error[E0308]: mismatched types
 --> src/lib.rs:8:3
  |
8 |   uwu::<A>(|| {});
  |   ^^^^^^^^^^^^^^^ types differ
  |
  = note: expected trait `Fn<A>`
             found trait `Fn()`
note: required by a bound in `uwu`
 --> src/lib.rs:5:26
  |
5 | fn uwu<A: Tuple>(_: impl Fn<A, Output = ()>) {}
  |                          ^^^^^^^^^^^^^^^^^^ required by this bound in `uwu`
```

I'd rather not delay a bug here and just let inference fail gracefully, since the type mismatch is reported *after* the closure is type-checked.